crates/proto/src/xfer/dnssec_dns_handle.rs

```diff
@@ -382,25 +382,28 @@ where
 {
     // wrapper for some of the type conversion for typed DNSKEY fn calls.
 
-    match rrset.record_type() {
-        // validation of DNSKEY records require different logic as they search for DS record coverage as well
-        RecordType::DNSKEY => verify_dnskey_rrset(handle.clone_with_context(), rrset, options)
-            .await
-            .map(|proof| (proof, None)),
-        _ => verify_default_rrset(&handle.clone_with_context(), rrset, rrsigs, options).await,
+    if matches!(rrset.record_type(), RecordType::DNSKEY) {
+        let is_trust_anchor =
+            pre_verify_dnskey_rrset(handle.clone_with_context(), &rrset, options).await?;
+
+        if is_trust_anchor {
+            return Ok((Proof::Secure, None));
+        }
     }
+
+    verify_default_rrset(&handle.clone_with_context(), rrset, rrsigs, options).await
 }
 
-/// Verifies a dnskey rrset
+/// Pre-verifies a dnskey rrset
 ///
 /// This first checks to see if the key is in the set of trust_anchors. If so then it's returned
-///  as a success. Otherwise, a query is sent to get the DS record, and the DNSKEY is validated
+///  as a trusted key. Otherwise, a query is sent to get the DS record, and the DNSKEY is validated
 ///  against the DS record.
-async fn verify_dnskey_rrset<H>(
+async fn pre_verify_dnskey_rrset<H>(
```

Comment:
Not sure this name and type change are actually helpful? All of this code seems to be pretty crappy in terms of describing the invariants/semantics...

Would be interesting if you can try to rewrite the docstring to more closely describe what the code is actually doing here, and how it fits in relative to `verify_default_rrset()`.

Reply:
Thanks!